README.md
# @topsort/building-blocks

![version](https://img.shields.io/npm/v/@topsort/building-blocks)
![downloads](https://img.shields.io/npm/dw/@topsort/building-blocks)
![license](https://img.shields.io/github/license/topsort/building-blocks)
![GitHub Repo stars](https://img.shields.io/github/stars/topsort/building-blocks?style=social)


> Topsort's JS library for building individual and easily integratable widgets.

## Table of Contents

1. [Integration](#integration)
   1. [Product Promotion Modal](#product-promotion-modal)
1. [Development](#development)
   1. [Running the demo](#running-the-demo)
   1. [Adding new environment variables](#adding-new-environment-variables)

## Integration

1. Insert a script in your site's `<head>` tag to load the Topsort Building Blocks library. Ensure the script has `defer` to load it without affecting your site load times. Replace `x.y.z` with the desired version, or use `@latest` to use the latest version of the library at all times.

   ```html
   <head>
     <!-- The rest -->
     <script
       defer
       src="https://unpkg.com/@topsort/building-blocks@x.y.z/dist/index.js"
     ></script>
   </head>
   ```

1. After the DOM is loaded, create an instance of the `TopsortBlocks` library. For example:

   ```js
   document.addEventListener("DOMContentLoaded", async () => {
     const tsBlocks = new TopsortBlocks();
   });
   ```

1. Initialize it:

   With defaults:

   ```js
   await tsBlocks.init({
     authUrl: "https://topsort.auth.mymarketplace.com",
     externalVendorId: "vendor-id-123",
   });
   ```

   With custom props:

   ```js
   await tsBlocks.init({
    // auth settings
     authUrl: "https://topsort.auth.mymarketplace.com",
     extraAuthHeaders: {
      Authorization: `Bearer api-key-123`,
    },
     externalVendorId: "vendor-id-123",
     // if you want to use a custom target class:
     promoteTargetClassName: "my-custom-promote-target",
     style: {
       // color RGBs acceptable in either format:
       primaryColorRgb: "50, 175, 200",
       secondaryColorRgb: "50, 175, 200",
       fontColorRgb: [50, 175, 200],

       button: {
         // defaults to "sm":
         borderRadius: "none" | "sm" | "full",
       },
     },
     text: {
       // defaults to "Promote":
       promoteButton: "Create Campaign",
       // defaults to "See Campaign":
       detailButton: "View Campaign",
     },
   });
   ```

### Where do I get authUrl from!?

##### TL;DR

The `authUrl` is the URL to your backend endpoint that requests the JWT from the Topsort Modal Auth API and returns its response. It acts as a proxy, enhancing security by keeping your advanced API key secure and enabling you to interact with Topsort's systems.

The Modal API employs JSON Web Tokens (JWT) to validate every request made to our systems. To acquire a JWT, you'll need to hit our public Modal Auth API, supplying the `vendorId` used by the library. In response, we'll provide a JWT that authorizes that specific vendor to perform tasks offered by the Modal library for a restricted time period.

Here's an example on how to request a JWT:

```bash
curl --request GET \
     --url https://api.topsort.com/public/v1/modal-service/auth/vendors/vendor-id \
     --header 'accept: application/json' \
     --header 'authorization: Bearer <YOUR_ADVANCED_API_KEY>'
```

Note that this request mandates the use of your `ADVANCED_API_KEY` as authorization.

##### Protecting Your API Key

Given that the API key allows campaign creation, deletion, catalog management, and more, it's crucial to prevent its exposure. We therefore advise our customers to establish an endpoint on their backend that makes this request.

This backend endpoint should be accessible via `authUrl`. In essence, the `authUrl` is the URL to the backend endpoint that will request the JWT from the Modal Auth API. Think of it as a proxy call!

By implementing this setup, you can safely and securely interact with Topsort's systems and manage your ad campaigns. You must set the `authUrl` during initialization. Optionally, you can pass headers for the request to your backend, as shown in the following example:

 ```js
await tsBlocks.init({
    // auth settings
    authUrl: "https://topsort.auth.mymarketplace.com",
    extraAuthHeaders: {
        Authorization: `Bearer api-key-123`,
    }
});
```

##### Summary

To put it simply, you need to make a request to `${authUrl}/auth/vendors/${vendorId}` on your backend, which in turn should call our Modal Auth API at `https://api.topsort.com/public/v1/modal-service/auth/vendors/${vendorId}`. Your endpoint will then return whatever response it receives from our API.

### Shop Promotion Button

To add a button that will create a campaign with all the vendor products, add the following HTML class to
the element you want the button appended to:

  ```html
  <div class="ts-promote-shop" data-ts-shop-name="My Shop"></div>
  ```

Then simply call `useShopPromotion`  to have the button rendered and functional, like so:

  ```javascript
  tsBlocks.useShopPromotion();
  ```

### Product Promotion Button and Modal (Optional)

If there is no product related markup added, the only option to promote a product would be through a 'Promote My Shop' button which will add all vendor products to a campaign.

If you decide to integrate the product promotion feature, follow these steps:

1. In your markup, add the following HTML class and data attributes to the element(s) you want a Promote button appended to:

   ```html
   <div
     class="ts-promote-target some-other-custom-class"
     data-ts-product-id="product-id-123"
     data-ts-product-name="My Product"
     data-ts-product-img-url="https://picsum.photos/100"
   >
     ...
   </div>
   ```

   This target class is also a static property of the TopsortBlocks class:

   ```jsx
   <div
     class={`${TopsortBlocks.promoteTargetClassName} my-custom-class`}
     data-ts-product-id="product-id-123"
     data-ts-product-name="My Product"
     data-ts-product-img-url="https://picsum.photos/100"
   >
     ...
   </div>
   ```

   If you are using a custom promote target class name, use it instead:

   ```html
   <div
     class="my-custom-promote-target my-custom-class"
     data-ts-product-id="product-id-123"
     data-ts-product-name="My Product"
     data-ts-product-img-url="https://picsum.photos/100"
   >
     ...
   </div>
   ```

1. Call `useProductPromotion` to have the Promote buttons rendered, and each time you want to have the Promote buttons rerendered such as after navigating to a new page of products or filtering the product list:

   ```js
   tsBlocks.useProductPromotion();
   ```

## Development

### Running the demo

1. Install the dependencies:

   ```bash
   pnpm install
   ```

1. Create and set up your `.env` file:

   ```bash
   cp .env.example .env
   ```

1. Start the development server. The console will output a clickable URL.

   ```bash
   pnpm run start
   ```

### Adding new environment variables

Environment variables should not contain secrets (e.g. private API keys) since they will be injected into the published build. They are meant to change the behaviour between different environments such as calling a local, staging, or production Central Services API base URL.

To add a new env var:

1. Add its name to `.env.example` with an empty or default value:

   ```bash
   MY_NEW_PUBLIC_VAR=
   ```

1. Add its name and value to your `.env`:

   ```bash
   MY_NEW_PUBLIC_VAR="the actual value"
   ```

1. To have it injected into the app, add it to `webpack.common.js` in the `DefinePlugin`:

   > Remember to wrap the value in `JSON.stringify()` because the `DefinePlugin` does a find-and-replace and the value given to it must include actual quotes inside of the string itself.
   >
   > For reference, see the Tip on the [Webpack DefinePlugin docs](https://webpack.js.org/plugins/define-plugin/).

   ```js
   new webpack.DefinePlugin({
    // ...the rest
    MY_NEW_PUBLIC_VAR: JSON.stringify(process.env.MY_NEW_PUBLIC_VAR),
   }),
   ```

1. To inform TypeScript of this new global variable, add it to `src/types.ts` under the global declaration with its type:

   ```ts
   declare global {
     // ...the rest
     const MY_NEW_PUBLIC_VAR: string;
   }
   ```

1. To inform eslint of this new global variable, add it to `.eslintrc.js` under `globals` with a `readonly` value (assuming it's meant to be readonly):

   ```js
   globals: {
    // ...the rest
    MY_NEW_PUBLIC_VAR: "readonly",
   },
   ```

1. Ensure you restart your dev server so webpack can pick up the latest changes.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is loosely based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
We follow the format used by [Open Telemetry](https://github.com/open-telemetry/opentelemetry-python/blob/main/CHANGELOG.md).

#### Version 2.1.0 (2023-09-04)

##### Build System / Dependencies

* **deps:**  bump aws-actions/configure-aws-credentials from 2 to 3 ([17e1f53b](https://github.com/Topsort/building-blocks/commit/17e1f53bd551f2e080ff756080ed20a10e678c74))
* **deps-dev:**
  *  bump the lint group with 3 updates ([236d11de](https://github.com/Topsort/building-blocks/commit/236d11debc1e928bb848e892db1f5ee05ba16a8e))
  *  bump typescript from 5.1.6 to 5.2.2 ([95b9c445](https://github.com/Topsort/building-blocks/commit/95b9c4454173dab58b374a31b7a1c70926b05c95))

##### Chores

*  switch user-agent key ([d528d2af](https://github.com/Topsort/building-blocks/commit/d528d2afd91b062230e6e45f1f8406df4c754dd8))

##### New Features

*  create chart component ([#162](https://github.com/Topsort/building-blocks/pull/162)) ([ea32f4bf](https://github.com/Topsort/building-blocks/commit/ea32f4bf9e4d792adca08a00d4b4e226e11a53c7))
*  date dropdown component ([#158](https://github.com/Topsort/building-blocks/pull/158)) ([223fc198](https://github.com/Topsort/building-blocks/commit/223fc198feba1115edf035b134c15d5781e8d818))

##### Bug Fixes

*  switch to npm for publishing ([44f34c36](https://github.com/Topsort/building-blocks/commit/44f34c3601f7e544fad625533c5673c8d2b39140))

##### Other Changes

*  switch to npm for publishing ([deaccd91](https://github.com/Topsort/building-blocks/commit/deaccd91a6238e679d62718d36bf599dacfe9865))

## Version 2.1.0-beta2 (2023-08-22)

### Added
- Added dropdown component (#153)
  ![image](https://github.com/Topsort/building-blocks/assets/95320456/2756830b-01fb-4ff6-a70b-0d91495e8243)

### Changed
- Updated reporting path in central services (#154)
- Improved HTML for better validation (#152)

### Fixed
- Fixed path to types and requests (#151)

### Chore
- Added badges to README (#149)
- Used environment variables instead of hardcoded values (#143)
- Added `CODEOWNERS` file for code ownership (#147)
- Updated dependencies: bumped `preact` from 10.17.0 to 10.17.1 (#146)
- Updated dependencies: bumped `zod` from 3.22.1 to 3.22.2 (#145)

## Version 2.1.0-beta1 (2023-08-03)

### Changed
- feat: Promote my Shop button. Add a new button that can create a campaign using the whole catalog of the vendor.
  * Create campaign
  * Obtain campaign
  * Update campaign


demo/loader.js
import { topsortedProducts } from "./topsortedProducts.js";

/* global TopsortBlocks */

// Demo config
const imgSize = 72;
const numProducts = 20;
const productsPerPage = 10;
const isUsingTopsortBlocks = true;
const isUsingCustomProps = false;
const useTopsortedProducts = true;
const customPromoteTargetClassName = "my-custom-promote-target";
const CENTRAL_SERVICES_URL = "https://api.topsort.com";
const AUTH_BASE_URL = "https://buildingblocks.api.topsort.ai";

function getNewElement(selector) {
  const proto = document.querySelector(selector);
  const newNode = proto.cloneNode(true);
  newNode.hidden = false;

  // Remove the id on the new element so that the only place
  // the id exists is on the original node for future cloning.
  newNode.id = "";

  return newNode;
}

function createProductElement(product) {
  const productElement = getNewElement("#product-proto");

  const img = productElement.querySelector(".product-img");
  img.src = product.imgUrl;
  img.width = imgSize;
  img.height = imgSize;

  productElement.querySelector(".product-name").innerText = product.name;
  productElement.querySelector(".product-code").innerText = product.id;
  productElement.querySelector(".product-quantity").innerText =
    product.quantity;
  productElement.querySelector(
    ".product-price"
  ).innerText = `$${product.price}`;
  productElement.querySelector(".product-status").innerText = "Active";

  const target = productElement.querySelector(".promote-target-placeholder");
  target.classList.remove("promote-target-placeholder");

  // return early if we should skip over this product for demo purposes
  if (product.skip) return productElement;

  target.classList.add(
    isUsingCustomProps
      ? customPromoteTargetClassName
      : TopsortBlocks.promoteTargetClassName
  );
  target.dataset.tsProductId = product.id;
  target.dataset.tsProductName = product.name;
  target.dataset.tsProductImgUrl = product.imgUrl;

  return productElement;
}

function getNumberedProducts() {
  const products = [];
  for (let i = 1; i < numProducts + 1; i++) {
    products.push({
      id: `product-${i}`,
      imgUrl: `https://picsum.photos/${imgSize}?random=${i}`,
      name: `Product ${i}`,
      price: `${i}.99`,
      quantity: i,
      // to demo NOT putting a promote button on a certain product
      skip: i === 8,
    });
  }
  return products;
}

document.addEventListener("DOMContentLoaded", async () => {
  if (window.TopsortBlocks) {
    console.log("[TopsortBlocks] Module loaded");
  } else {
    console.error("[TopsortBlocks] Module did not load correctly.");
    return;
  }

  const wrapper = document.querySelector(".wrapper");

  const products = useTopsortedProducts
    ? topsortedProducts
    : getNumberedProducts();

  let filter = "";
  let page = 1;

  const pageSelect = document.getElementById("page-select");
  const searchInput = document.getElementById("product-search");

  const updateProductElements = () => {
    document.querySelectorAll(".product").forEach((productElement) => {
      if (productElement.id !== "product-proto") {
        wrapper.removeChild(productElement);
      }
    });

    // filter products by name/id based on the entered filter
    const filteredProducts = products.filter(
      (product) =>
        product.name.toLowerCase().includes(filter) ||
        product.id.toLowerCase().includes(filter)
    );

    // it indicates the starting index of the products to be displayed
    const offset = (page - 1) * productsPerPage;

    // create product elements
    const productElements = filteredProducts
      .slice(offset, offset + productsPerPage)
      .map((product) => createProductElement(product));
    productElements.forEach((productElement) => {
      wrapper.appendChild(productElement);
    });

    // update the page numbers in select component
    // first remove all page numbers
    // then add new numbers based on the number of products
    pageSelect.querySelectorAll("option").forEach((option) => {
      pageSelect.removeChild(option);
    });

    [
      ...Array(Math.ceil(filteredProducts.length / productsPerPage)).keys(),
    ].forEach((pageIndex) => {
      const option = document.createElement("option");
      option.value = pageIndex + 1;
      option.text = pageIndex + 1;
      pageSelect.appendChild(option);
    });
    pageSelect.value = page;
  };

  updateProductElements();

  searchInput.addEventListener("input", (event) => {
    filter = event.target.value.toLowerCase();
    page = 1;
    updateProductElements();
    if (isUsingTopsortBlocks) {
      tsBlocks.useProductPromotion();
    }
  });

  pageSelect.addEventListener("change", (event) => {
    page = parseInt(event.target.value, 10);
    updateProductElements();
    if (isUsingTopsortBlocks) {
      tsBlocks.useProductPromotion();
    }
  });

  if (!isUsingTopsortBlocks) return;

  const tsBlocks = new TopsortBlocks();
  await tsBlocks.init({
    // This helper calls our demo marketplace backend to retrieve the authToken.
    // You can manually set an auth token here for development purposes.
    centralServicesUrl: `${CENTRAL_SERVICES_URL}`,
    authUrl: `${AUTH_BASE_URL}`,
    extraAuthHeaders: {
      Authorization: `Bearer api-key-123`,
    },
    externalVendorId: "wendys",
    ...(isUsingCustomProps && {
      promoteTargetClassName: customPromoteTargetClassName,
      style: {
        primaryColorRgb: "120, 170, 50",
        secondaryColorRgb: "55, 150, 220",
        fontColorRgb: [200, 80, 215],
        button: {
          borderRadius: "none",
        },
      },
      text: {
        promoteButton: "Create Campaign",
        detailButton: "View Campaign",
      },
    }),
  });

  tsBlocks.useProductPromotion();
  tsBlocks.useShopPromotion();
});
